src/openms/source/ANALYSIS/ID/PrecursorPurity.cpp

```diff
@@ -0,0 +1,231 @@
+
+    auto lower_it = ms1.MZBegin(lower);
+    auto upper_it = ms1.MZEnd(upper);
+
+    // std::cout << "charge: " << charge << " | lower: " << lower << " | target: " << target_mz << " | upper: " << upper << std::endl;
+    // std::cout << "lower offset: " << pre.getIsolationWindowLowerOffset() << " | upper offset: " << pre.getIsolationWindowUpperOffset() << std::endl;
+    // std::cout << "lower peak: " << (*lower_it).getMZ() << " | upper peak: " << (*upper_it).getMZ() << std::endl;
+
+    PeakSpectrum isolated_window;
+    while (lower_it != upper_it)
+    {
+      isolated_window.push_back(*lower_it);
+      lower_it++;
+    }
+
+    // std::cout << "Isolation window peaks: " << isolated_window.size();
+    for (auto peak : isolated_window)
+    {
+      // std::cout << " | " << peak.getMZ();
+    }
+    // std::cout << std::endl;
+
+    // total intensity in isolation window
+    double total_intensity(0);
+    double target_intensity(0);
+    Size target_peak_count(0);
+    for (auto peak : isolated_window)
+    {
+      total_intensity += peak.getIntensity();
+    }
+
+    // search for the target peak, return scores with 0-values if it is not found
+    if (isolated_window.empty())
+    {
+      return score;
+    }
+
+    // estimate a lower boundary for isotopic peaks
+    int negative_isotopes((pre.getIsolationWindowLowerOffset() * charge));
+    double iso = -negative_isotopes;
+    // depending on the isolation window, the first estimated peak might be outside the window
+    if (target_mz + (iso * Constants::C13C12_MASSDIFF_U / charge) < lower)
+    {
+      iso++;
+    }
+
+    // std::cout << "target peaks: ";
+    // deisotoping (try to find isotopic peaks of the precursor mass, even if the actual precursor peak is missing)
+    while (true) // runs as long as the next mz is within the isolation window
+    {
+      double next_peak = target_mz + (iso * Constants::C13C12_MASSDIFF_U / charge);
+      // std::cout << iso << " : iso | " << next_peak << " : next peak | ";
+
+      // stop loop when new mz is outside the isolation window
+      // changes through the isotope index iso
+      if (next_peak > upper)
+      {
+        break;
+      }
+      int next_iso_index = isolated_window.findNearest(next_peak, precursor_tolerance_abs);
+      if (next_iso_index != -1)
+      {
+        target_intensity += isolated_window[next_iso_index].getIntensity();
+
+        // std::cout << isolated_window[next_iso_index].getMZ() << " : matched | ";
+
+        isolated_window.erase(isolated_window.begin()+next_iso_index);
+        target_peak_count++;
+      }
+      // always increment iso to progress the loop
+      iso++;
+    }
+    // std::cout << std::endl;
+
+    // std::cout << "noise peaks: ";
+    double noise_intensity(0);
+    for (auto peak : isolated_window)
+    {
+      noise_intensity += peak.getIntensity();
```

Comment:
this is incorrect, as you erased all noise on line 117 except the noise between the last isotopic peak and the end